src/Control/Distributed/Process.hs

```diff
@@ -402,12 +407,22 @@ spawnChannelLocal proc = do
 callLocal ::
      Process a  -- ^ Process to run
   -> Process a  -- ^ Value returned
-callLocal proc = do
-  mv <- liftIO newEmptyMVar
-  self <- getSelfPid
-  _ <- spawnLocal $ do
-     link self
-     try proc >>= liftIO . putMVar mv
-  liftIO $ takeMVar mv >>=
-     either (throwIO :: SomeException -> IO a) return
-
+callLocal proc = mask $ \release -> do
+  mv    <- liftIO newEmptyMVar :: Process (MVar (Either SomeException a))
+  child <- spawnLocal $ do
+             ep <- try $ release $ proc
+             liftIO $ putMVar mv ep
+  liftIO (fetchResult mv) `onException`
+    (do exit child "exception in parent process"
+        waitForExit child)
+  where
+    waitForExit child = do
+      bracket (monitor child)
+              (unmonitor)
+              (\mRef -> receiveWait
+                   [ matchIf (\(ProcessMonitorNotification ref _ _) -> ref == mRef)
+                             (\_ -> return ())
+                   ])
+    fetchResult mv = do
+        rs <- takeMVar mv
+        liftIO $ either throwIO return rs
```

Comment:
This call to `liftIO` seems to be redundant. It would be more conspicuous to inline `fetchResult` at the call site.